misk-redis/src/main/kotlin/misk/redis/RealPipelinedRedis.kt

```diff
@@ -0,0 +1,576 @@
+  }
+
+  override fun hget(key: String, field: String): Supplier<ByteString?> {
+    val keyBytes = key.toByteArray(charset)
+    val fieldBytes = field.toByteArray(charset)
+    val response = pipeline.hget(keyBytes, fieldBytes)
+    return Supplier { response.get()?.toByteString() }
+  }
+
+  override fun hgetAll(key: String): Supplier<Map<String, ByteString>?> {
+    val keyBytes = key.toByteArray(charset)
+    val response = pipeline.hgetAll(keyBytes)
+    return Supplier {
+      response.get()
+        ?.mapKeys { it.key.toString(charset) }
+        ?.mapValues { it.value.toByteString() }
+    }
+  }
+
+  override fun hlen(key: String): Supplier<Long> {
+    val keyBytes = key.toByteArray(charset)
+    val response = pipeline.hlen(keyBytes)
+    return Supplier { response.get() }
+  }
+
+  override fun hmget(key: String, vararg fields: String): Supplier<List<ByteString?>> {
+    val keyBytes = key.toByteArray(charset)
+    val fieldsAsByteArrays = fields.map { it.toByteArray(charset) }.toTypedArray()
+    val response = pipeline.hmget(keyBytes, *fieldsAsByteArrays)
+    return Supplier { response.get()?.map { it?.toByteString() } ?: emptyList() }
+  }
+
+  override fun hincrBy(key: String, field: String, increment: Long): Supplier<Long> {
+    val keyBytes = key.toByteArray(charset)
+    val fieldBytes = field.toByteArray(charset)
+    val response = pipeline.hincrBy(keyBytes, fieldBytes, increment)
+    return Supplier { response.get() }
+  }
+
+  override fun hrandFieldWithValues(key: String, count: Long): Supplier<Map<String, ByteString>?> {
+    checkHrandFieldCount(count)
+    val keyBytes = key.toByteArray(charset)
+    val response = pipeline.hrandfieldWithValues(keyBytes, count)
+    return Supplier {
+      response.get()
+        ?.mapKeys { it.key.toString(charset) }
+        ?.mapValues { it.value.toByteString() }
+    }
+  }
+
+  override fun hrandField(key: String, count: Long): Supplier<List<String>> {
+    checkHrandFieldCount(count)
+    val keyBytes = key.toByteArray(charset)
+    val response = pipeline.hrandfield(keyBytes, count)
+    return Supplier { response.get()?.map { it.toString(charset) } ?: emptyList() }
+  }
+
+  override fun set(key: String, value: ByteString, expiryDuration: Duration?): Supplier<Unit> {
+    val keyBytes = key.toByteArray(charset)
+    val valueBytes = value.toByteArray()
+    val params = SetParams().apply {
+      if (expiryDuration != null) {
+        px(expiryDuration.toMillis())
+      }
+    }
+    val response = pipeline.set(keyBytes, valueBytes, params)
+    return Supplier { response.get() }
+  }
+
+  override fun setnx(key: String, value: ByteString, expiryDuration: Duration?): Supplier<Boolean> {
+    val keyBytes = key.toByteArray(charset)
+    val valueBytes = value.toByteArray()
+    val params = SetParams().nx().apply {
+      if (expiryDuration != null) px(expiryDuration.toMillis())
+    }
+    val response = pipeline.set(keyBytes, valueBytes, params)
```

Comment:
For `pipelining`, `setnx` is actually backed by `SET` with `NX` and optional `PX` arguments. This is because `SETNX` is deprecated as of Redis 2.6 and I don't want to support this with pipelining.

We should follow up and change the [implementation](https://github.com/cashapp/misk/blob/master/misk-redis/src/main/kotlin/misk/redis/RealRedis.kt#L192) of the `setnx` method without expiry durations to _also_ be backed by `SET` with the `NX` parameters.